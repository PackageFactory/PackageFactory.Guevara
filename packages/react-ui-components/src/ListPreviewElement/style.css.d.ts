export interface IStyleCss {
  'listPreviewElement': string;
  'listPreviewElement--isDisabled': string;
  'listPreviewElement--isHighlighted': string;
  'listPreviewElement__icon': string;
}

export const locals: IStyleCss;
