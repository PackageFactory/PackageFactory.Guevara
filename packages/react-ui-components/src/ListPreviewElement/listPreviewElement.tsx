import mergeClassNames from 'classnames';
import React, {PureComponent, ReactNode} from 'react';

import Icon from '../Icon';
import {PickDefaultProps} from '@neos-project/react-ui-components/types';

export interface ListPreviewElementProps {
    // ------------------------------
    // API inside custom ListPreviewElements
    // ------------------------------
    readonly icon?: string;
    readonly className?: string;
    readonly disabled?: boolean;
    readonly isHighlighted?: boolean;
    readonly onClick?: () => void;
    readonly onMouseEnter?: () => void;
    readonly children?: ReactNode;

    // ------------------------------
    // Theme & Dependencies
    // ------------------------------
    readonly theme?: ListPreviewElementTheme;
}

interface ListPreviewElementTheme {
    readonly 'listPreviewElement': string;
    readonly 'listPreviewElement--isHighlighted': string;
    readonly 'listPreviewElement--isDisabled': string;
    readonly 'listPreviewElement__icon': string;
}

const defaultProps: PickDefaultProps<ListPreviewElementProps, 'isHighlighted'> = {
    isHighlighted: false,
};

/**
 * The ListPreviewElement is responsible for rendering a single element in a Select Box
 * or a MultiSelectBox.
 *
 * It encapsulates basic styling and functionality needed by the SelectBox / MultiSelectBox.
 *
 * **Instead of directly using this component, you might want to use SelectBox_Option_SingleLine
 * or SelectBox_Option_MultiLineWithThumbnail instead, as these provide default styling.**
 *
 * Often, you will create your own ListPreviewElements, taking this element as a basis like the following:
 * ```
 * const MySpecialPreviewElement = props => {
 *      return (
 *          <ListPreviewElement {...props} icon={props.yourIconHere}>
 *               ... your content here ...
 *          </ListPreviewElement>
 *      );
 * }
 * ```
 */
export default class ListPreviewElement extends PureComponent<ListPreviewElementProps> {
    public static readonly defaultProps = defaultProps;

    public render(): JSX.Element {
        const {
            icon,
            className,
            disabled,
            children,

            isHighlighted,

            theme,
        } = this.props;

        const optionClassName = mergeClassNames(
            theme!.listPreviewElement,
            {
                [theme!['listPreviewElement--isHighlighted']]: isHighlighted,
                [theme!['listPreviewElement--isDisabled']]: disabled,
            },
            className
        );

        return (
            <div
                onMouseEnter={this.handleMouseEnter}
                onClick={this.handleClick}
                className={optionClassName}
                role="button"
            >
                {Boolean(icon) && <Icon className={theme!.listPreviewElement__icon} icon={icon!}/>}
                {children}
            </div>
        );
    }

    private readonly handleMouseEnter = () => {
        if (!this.props.disabled && this.props.onMouseEnter) {
            this.props.onMouseEnter();
        }
    }

    private readonly handleClick = () => {
        if (!this.props.disabled && this.props.onClick) {
            this.props.onClick();
        }
    }
}
