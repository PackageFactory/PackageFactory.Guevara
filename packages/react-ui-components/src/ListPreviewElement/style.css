.listPreviewElement {
    composes: reset from './../reset.css';
    background: var(--brandColorsContrastNeutral);
    cursor: pointer;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    font-size: 14px;

    /* subtract the size of padding */
    line-height: calc(var(--goldenUnit) - 10px);
    padding: 5px var(--spacing) !important;
}

.listPreviewElement--isHighlighted {
    background: var(--brandColorsPrimaryBlue);
}

.listPreviewElement__icon {
    composes: reset from './../reset.css';
    padding-right: .5em;
}
