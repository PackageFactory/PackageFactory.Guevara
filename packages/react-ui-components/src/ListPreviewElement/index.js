import {themr} from '@friendsofreactjs/react-css-themr';
import identifiers from './../identifiers';
import style from './style.css';
import ListPreviewElement from './listPreviewElement';

const ThemedListPreviewElement = themr(identifiers.listPreviewElement, style)(ListPreviewElement);

//
// Dependency injection
//
import injectProps from './../_lib/injectProps.js';
import Icon from './../Icon/index';

export default injectProps({
    Icon
})(ThemedListPreviewElement);

