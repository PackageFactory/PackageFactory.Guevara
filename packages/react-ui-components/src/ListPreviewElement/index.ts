import {themr} from '@friendsofreactjs/react-css-themr';

import identifiers from '../identifiers';
import ListPreviewElement from './listPreviewElement';
import style from './style.css';

export default themr(identifiers.listPreviewElement, style)(ListPreviewElement);
