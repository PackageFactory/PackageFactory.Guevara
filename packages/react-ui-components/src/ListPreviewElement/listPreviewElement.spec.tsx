import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import ListPreviewElement from '.';
import {ListPreviewElementProps} from './listPreviewElement';

describe('ListPreviewElement', () => {
    const props: ListPreviewElementProps = {
        children: 'Foo Children',
        theme: {
            listPreviewElement: 'listPreviewElementClassName',
            'listPreviewElement--isDisabled': 'disabledClassName',
            'listPreviewElement--isHighlighted': 'highlightedClassName',
            listPreviewElement__icon: 'iconClassName',
        }
    };

    it('should render correctly.', () => {
        const wrapper = shallow(<ListPreviewElement {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should allow the propagation of "className" with the "className" prop.', () => {
        const wrapper = shallow(<ListPreviewElement {...props} className="fooClassName"/>);

        expect(wrapper.prop('className')).toContain('fooClassName');
    });

    it('should allow the propagation of additional props to the wrapper.', () => {
        const wrapper = shallow(<ListPreviewElement {...props} data-foo="bar"/>);

        expect(wrapper.prop('data-foo')).toBe('bar');
    });

    it('should handle the click callback if the component is clicked.', () => {
        const onClick = jest.fn();
        const wrapper = shallow(<ListPreviewElement {...props} onClick={onClick}/>);

        wrapper.simulate('click');

        expect(onClick.mock.calls.length).toBe(1);
    });
});
