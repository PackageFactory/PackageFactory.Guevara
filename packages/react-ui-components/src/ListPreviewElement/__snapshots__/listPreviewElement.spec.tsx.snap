// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPreviewElement should render correctly. 1`] = `
<ListPreviewElement
  isHighlighted={false}
  theme={
    Object {
      "listPreviewElement": "listPreviewElementClassName",
      "listPreviewElement--isDisabled": "disabledClassName",
      "listPreviewElement--isHighlighted": "highlightedClassName",
      "listPreviewElement__icon": "iconClassName",
    }
  }
>
  Foo Children
</ListPreviewElement>
`;
